### DISCUSSION TITLE
Event processing

### DISCUSSION THREAD

**Discussion by afflom** [emporous/community#5] [Design]
Hi all, 

These drawings depict the event workflow for a UOR Collection. 

## Event Engine Lookup
The first drawing depicts Event Engine lookup. When a user(caller) "runs" a collection, the first task performed by the client is the Event Engine lookup. Event Engines are imported into Collections via Schema Collections. The Client first looks for the Schema root node in the addressed Collection. The Client then retrieves the Schema Root node and discovers the Event Engine root node. The Client then retrieves the discovered Event Engine root node and then discovers the Event Engine for the requested platform and architecture. The Client then downloads the Event Engine. 

![image](https://user-images.githubusercontent.com/30676559/187067722-8c795d82-7c0b-42a1-8161-5d644bd815ae.png)

## Event Engine Entrypoint Processing
After the Event Engine lookup has completed, the Event Engine is executed in an isolated environment. The UOR Client combines user input with the attributes of the Entrypoint discovered from the imported schema and sends this as attributes to the Event Engine. If an Entrypoint includes a reference to content, a payload is sent to the Event Engine as well. The Event Engine begins processing the Entrypoint when it is received. 

![image](https://user-images.githubusercontent.com/30676559/187068285-072d9daf-d7d0-41fa-89bf-a3516838146f.png)

## Event Engine Processing
After the Event Engine begins processing its Entrypoint, it may be instructed to retrieve additional content from the referenced Collection. When this happens, the Event Engine requests content from the Client. The Client then discovers the requested content and then returns it to the Event Engine for processing. 
 
![image](https://user-images.githubusercontent.com/30676559/187068737-408944ee-57ae-4414-b03a-d6fdc2f34bbd.png)

## Return
After the Event Engine has finished processing, it returns any requested output with an exit code and is then terminated. 

## Linked Collections
When a Linked Collection is called by an Event Engine, the calling Event Engine becomes the caller and the Event Process starts over for the referenced Linked Collection. Once finished, the Linked Collection Event Engine returns the requested output to the calling Collection and the calling Collection resumes its processing. 

Please let me know if you have any questions or comments.

Related: #4 and #6

**Comment by jpower432** [comment]
I think @usrbinkat's thoughts would be great based on the discussion from #3. If we need compatibility with legacy applications using a mutating webhook, maybe we need an admission controller in this workflow.

**Reply by usrbinkat** [reply]
+1 for admission controller

**Reply by jpower432** [reply]
Another thought for the admission controller is to protect against issues that can arise with mutably tagged references.
